<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2021-10-03" />

<title>RLBase</title>

<script src="documentation_files/header-attrs-2.11/header-attrs.js"></script>
<script src="documentation_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="documentation_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="documentation_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="documentation_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="documentation_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="documentation_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="documentation_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="documentation_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="documentation_files/navigation-1.1/tabsets.js"></script>
<link href="documentation_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="documentation_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore"><em>RLBase</em></h1>
<h3 class="subtitle">Documentation and FAQs</h3>
<h4 class="date">2021-10-03</h4>

</div>


<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>R-loops are three-stranded nucleic acid structures formed from the hybridization of RNA and DNA. With the advent of DNA-RNA immunoprecipitation sequencing (DRIP-Seq) in 2012, it become possible to map R-loop locations throughout the genome <span class="citation">(Ginno et al. 2012)</span>. To date, over 40 such studies have been performed, yielding hundreds of R-loop maps across a variety of species. <strong>However, inconsistent data quality has thus far thwarted every attempt to use meta-analysis to draw deeper insights about R-loop biology from this trove of data.</strong></p>
<div class="figure">
<img src="R-loop_promoting_factors.jpg" alt="Depiction of an R-loop and the factors which may promote them (Hegazy et al., 2020; PMID: [31843970](https://pubmed.ncbi.nlm.nih.gov/31843970/))." width="640" />
<p class="caption">
Depiction of an R-loop and the factors which may promote them (Hegazy et al., 2020; PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/31843970/">31843970</a>).
</p>
</div>
<p><strong>To overcome this limitation, we developed <a href="link_to_rlsuite.com">RLSuite</a>, a suite of tools for R-loop analysis with built-in quality control</strong>. RLSuite includes <a href="https://github.com/Bishop-Laboratory/RLPipes">RLPipes</a> (a CLI pipeline for upstream R-loop data processing), <a href="https://github.com/Bishop-Laboratory/RLSeq">RLSeq</a> (R package for downstream R-loop data analysis), and <a href="https://github.com/Bishop-Laboratory/RLBase">RLBase</a> (a web server and database to explore the results).</p>
<p><img src="RLBase_logo.png" align="right" style="border: none; float: right;" width="256" /></p>
<p><em>RLBase</em> is the first web server and database that provides un-restricted access to <strong>standardized and quality-controlled R-loop mapping data from hundreds of publicly-available samples</strong>. Moreover, process of curating these data has allowed us to confidently identify <strong>R-loop regions (RL Regions), sites of consensus R-loop formation</strong>. The RLBase web server is designed to enable access to all datasets along with the tools necessary to analyze them and evaluate user-supplied data.</p>
<p>Our goal in this document is to <strong>(1) showcase the intended usage of RLBase and (2) address FAQs which may arise when using RLBase</strong>.</p>
<details>
<summary>
<strong>Terminology explained</strong>
</summary>
<p>In this subsection, we describe the terminology used in <em>RLBase</em> to help aid in its usage.</p>
<h3>
Modes
</h3>
<p>RLBase contains R-loop mapping data from high-throughput sequencing experiments. There are, to date, over 20 different methods of R-loop mapping included in <em>RLBase</em>. To help summarize them, we have provided the following visualization:</p>
<div class="figure">
<img src="R-loop%20mapping%20methods%20(primary%20div)@2x.png" alt="Diagram showing the different R-loop mapping methods which exist. All of these are found in RLBase (except computational methods, SMRF-Seq, and bisDRIP-Seq)." width="1641" />
<p class="caption">
Diagram showing the different R-loop mapping methods which exist. All of these are found in RLBase (except computational methods, SMRF-Seq, and bisDRIP-Seq).
</p>
</div>
</details>
</div>
<div id="usage" class="section level1">
<h1>Usage</h1>
<p>The core features we intend <em>RLBase</em> to be used for are:</p>
<ul>
<li>Exploration of R-loop mapping samples (<a href="#rlsamples">view</a>)
<ul>
<li>Exploration of mapping samples from different modalities</li>
<li>Analysis of sample quality</li>
<li>Analysis of genomic feature enrichment</li>
</ul></li>
<li>R-loop Region (RL Region) analysis (<a href="#rlregions">link</a>)
<ul>
<li>Exploration of R-loop regions</li>
<li>Analysis of R-loop region correlation with gene expression</li>
</ul></li>
<li>User-supplied samples (<a href="#userdata">view</a>)
<ul>
<li>Analysis of user-supplied data</li>
</ul></li>
</ul>
<div id="rlsamples" class="section level2">
<h2>Analysis of RLBase samples</h2>
<p>To view samples in the RLBase database, navigate to the “Samples” tab.</p>
<details>
<summary>
<strong>Sample-page components explained</strong>
</summary>
<p>In this subsection, we detail the elements of the sample page to help elucidate its usage.</p>
<div class="figure">
<img src="R-loop_promoting_factors.jpg" alt="Sample page with " width="640" />
<p class="caption">
Sample page with
</p>
</div>
</details>
</div>
<div id="rlregions" class="section level2">
<h2>Exploration of R-loop regions</h2>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Ginno2012" class="csl-entry">
Ginno, Paul A., Paul L. Lott, Holly C. Christensen, Ian Korf, and Frédéric Chédin. 2012. <span>“R-Loop Formation Is a Distinctive Characteristic of Unmethylated Human CpG Island Promoters.”</span> <em>Molecular Cell</em> 45 (6): 814–25. <a href="https://doi.org/10.1016/j.molcel.2012.01.017">https://doi.org/10.1016/j.molcel.2012.01.017</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
